dev/test/bulk-writer.ts

```diff
@@ -0,0 +1,502 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {expect} from 'chai';
+import {Status} from 'google-gax';
+
+import {Firestore, setLogFunction, Timestamp, WriteResult} from '../src';
+import {BulkWriter} from '../src/bulk-writer';
+import {Deferred} from '../src/util';
+import {
+  ApiOverride,
+  create,
+  createInstance,
+  document,
+  remove,
+  response,
+  set,
+  update,
+  updateMask,
+  verifyInstance,
+} from './util/helpers';
+
+// Change the argument to 'console.log' to enable debug output.
+setLogFunction(() => {});
+
+const PROJECT_ID = 'test-project';
+
+interface BulkWriterRequest {
+  writes: Array<{[key: string]: {} | undefined}>;
+}
+
+interface BulkWriterResponse {
+  writeResults: Array<{[key: string]: {} | null}>;
+  status: Array<{[key: string]: {}}>;
+}
+
+interface RequestResponse {
+  request: BulkWriterRequest;
+  response: BulkWriterResponse;
+}
+
+describe('BulkWriter', () => {
+  let firestore: Firestore;
+  let bulkWriter: BulkWriter;
+  let requestCounter: number;
+  let opCount: number;
+  const flushDeferred = new Deferred<void>();
+  let flushCounter = 0;
+
+  beforeEach(() => {
+    requestCounter = 0;
+    opCount = 0;
+  });
+
+  function incrementOpCount(): void {
+    opCount++;
+  }
+
+  function verifyOpCount(expected: number): void {
+    expect(opCount).to.equal(expected);
+  }
+
+  function setOp(doc: string, value: string): BulkWriterRequest {
+    return set({
+      document: document(doc, 'foo', value),
+    }) as BulkWriterRequest;
+  }
+
+  function createOp(doc: string, value: string): BulkWriterRequest {
+    return create({
+      document: document(doc, 'foo', value),
+    }) as BulkWriterRequest;
+  }
+
+  function updateOp(doc: string, value: string): BulkWriterRequest {
+    const updateRequest = update({
+      document: document(doc, 'foo', value),
+      mask: updateMask('foo'),
+    }) as BulkWriterRequest;
+    delete updateRequest.writes[0].currentDocument;
```

Comment:
I don't quite understand why we are dropping the Precondition. Can you add a comment?